x-pack/legacy/plugins/maps/public/connected_components/map/mb/utils.js

```diff
@@ -11,27 +11,32 @@ export function removeOrphanedSourcesAndLayers(mbMap, layerList) {
 
   const mbStyle = mbMap.getStyle();
   const mbSourcesToRemove = [];
-  for (const sourceId in mbStyle.sources) {
-    if (mbStyle.sources.hasOwnProperty(sourceId)) {
+  for (const mbSourceId in mbStyle.sources) {
+    if (mbStyle.sources.hasOwnProperty(mbSourceId)) {
       const layer = layerList.find(layer => {
-        return layer.ownsMbSourceId(sourceId);
+        return layer.ownsMbSourceId(mbSourceId);
       });
       if (!layer) {
-        mbSourcesToRemove.push(sourceId);
+        mbSourcesToRemove.push(mbSourceId);
       }
     }
   }
-  const mbLayersToRemove = [];
-  mbStyle.layers.forEach(layer => {
-    if (mbSourcesToRemove.indexOf(layer.source) >= 0) {
-      mbLayersToRemove.push(layer.id);
+
+  const mbLayerIdsToRemove = [];
+  mbStyle.layers.forEach(mbLayer => {
+    const layer = layerList.find(layer => {
+      return layer.ownsMbLayerId(mbLayer.id);
+    });
+    if (!layer) {
+      mbLayerIdsToRemove.push(mbLayer.id);
     }
   });
-  mbLayersToRemove.forEach((layerId) => {
-    mbMap.removeLayer(layerId);
+
+  mbLayerIdsToRemove.forEach((mbLayerId) => {
+    mbMap.removeLayer(mbLayerId);
   });
-  mbSourcesToRemove.forEach(sourceId => {
-    mbMap.removeSource(sourceId);
+  mbSourcesToRemove.forEach(mbSourceId => {
```

Comment:
Might make sense to co-locate this `forEach` block right after the first section where the array is built

Reply:
I'll move that array building block down here, instead of moving this one up. Sources need to be removed _after_ layers are removed, since layers reference sources.